### PR TITLE
Add fold and unfold for BusinessCategoryDetails

### PR DESCRIPTION
**Pull request by cdepillabout**
I added a fold and unfold function for business category details.

@arowM: What do you think?

Pro:

- When adding a new data constructor to `BusinessCategoryDetail`, we don't have to worry about updating all of `toJSON`, `fromJSON`, `show`, `read`, etc.  Instead, we only have to update `foldBusinessCategoryDetail` and `unfoldBusinessCategoryDetail`.  (Well, then we will have to fix any compiler errors.)  This makes it slightly safer.

Con:

- The type signatures for some of the new functions are pretty complicated.  For example, `unfoldAllBusinessCategoryDetail` is pretty complicated.
- I had to use a `Proxy` in some of the functions.  It's kind of ugly.  I wanted to use ghc-8's visible type application, but I couldn't get it to work :-\

(This is for #65.)

### CODE REVIEW

File: src/Kucipong/Db/Models/Base.hs [hunk 1]
```diff
@@ -241,26 +266,194 @@ data BusinessCategoryDetail
...
+-- Helpful to use with functions like 'readMay'.
+--
+-- >>> let combiner = (<|>) :: Maybe z -> Maybe z -> Maybe z
+-- >>> let readProxy = Proxy :: Proxy Read
+-- >>> let f = readMay :: Read x => Text -> Maybe x
+-- >>> unfoldAllBusinessCategoryDetail readProxy combiner f "TravelingAsia"
+-- Just TravelingAsia
+-- >>> unfoldAllBusinessCategoryDetail readProxy combiner f "foobar"
+-- Nothing
+unfoldAllBusinessCategoryDetail
```
Comment (cdepillabout) [inline src/Kucipong/Db/Models/Base.hs:405]: The type signature of this function is pretty complicated.

File: src/Kucipong/Db/Models/Base.hs [hunk 2]
```diff
@@ -241,26 +266,194 @@ data BusinessCategoryDetail
...
+foldAllBusinessCategoryDetail
+  :: forall (c :: * -> Constraint) a proxy.
+     ( c GourmetDetail
+     , c FashionDetail
+     , c GadgetDetail
+     , c TravelingDetail
+     , c BeautyDetail
+     , c CommonDetail
+     )
+  => proxy c -> (forall x . c x => x -> a) -> BusinessCategoryDetail -> a
```
Comment (cdepillabout) [inline src/Kucipong/Db/Models/Base.hs:342]: I couldn't figure out a way to write this without using a `proxy`.

File: src/Kucipong/Db/Models/Base.hs [hunk 3]
```diff
@@ -275,30 +468,22 @@ instance PersistFieldSql BusinessCategoryDetail where
...
     where
+      mdetail :: Maybe BusinessCategoryDetail
       mdetail =
-        (GourmetDetail <$> parseMaybe parseJSON x) <|>
-        (FashionDetail <$> parseMaybe parseJSON x) <|>
-        (GadgetDetail <$> parseMaybe parseJSON x) <|>
-        (TravelingDetail <$> parseMaybe parseJSON x) <|>
-        (BeautyDetail <$> parseMaybe parseJSON x) <|>
-        (CommonDetail <$> parseMaybe parseJSON x)
+        unfoldAllBusinessCategoryDetailAlt
```
Comment (cdepillabout) [inline src/Kucipong/Db/Models/Base.hs:483]: Using `unfoldAllBusinessCategoryDetailAlt` makes this is slightly easier to write.  We don't have to worry about missing a case.

For example, if a new constructor `SportsDetail` was added to `BusinessCategoryDetail`, we would have to remember to update `parseJSON`, `businessCategoryDetailFromText`, etc.  However, if we use `unfoldAllBusinessCategoryDetailAlt`, all we have to remember is to update the `unfoldBusinessCategory` function

**Comment by arowM** [comment]
@cdepillabout 
It looks good to me, except that it may be a bit complicated for pros.
I'll completely agree with this way if we only have to update `BusinessCategoryDetail` type.
A one way to do so would be making compiler to automatically generate `unfoldAllBusinessCategoryDetailAlt` from `BusinessCateogryDetail` type with `TemplateHaskell`, `Data.Typeable`, etc.

**Comment by cdepillabout** [comment]
> I'll completely agree with this way if we only have to update BusinessCategoryDetail type.

I agree with this.

> A one way to do so would be making compiler to automatically generate `unfoldAllBusinessCategoryDetailAlt` from `BusinessCateogryDetail` type with `TemplateHaskell`, `Data.Typeable`, etc.

This is a really good idea.

I'll go ahead and merge this PR in, and then we can add this as a low-priority issue.

**Pull request merged**